pkg/kubectl/cmd/exec.go

```diff
@@ -174,16 +174,25 @@ func (p *ExecOptions) Run() error {
 		return err
 	}
 
-	if pod.Status.Phase != api.PodRunning {
-		return fmt.Errorf("pod %s is not running and cannot execute commands; current phase is %s", p.PodName, pod.Status.Phase)
-	}
-
 	containerName := p.ContainerName
 	if len(containerName) == 0 {
 		glog.V(4).Infof("defaulting container name to %s", pod.Spec.Containers[0].Name)
 		containerName = pod.Spec.Containers[0].Name
 	}
 
+	exists, isInitContainer := podHasContainerWithName(pod, containerName)
+	if !exists {
+		return fmt.Errorf("container %s is not valid for pod %s", containerName, p.PodName)
+	}
+
+	if !isInitContainer && pod.Status.Phase != api.PodRunning {
```

Comment:
I don't think you should be printing these errors here - this is the job of the server, and if the server isn't returning good errors, we should fix that.